File: ts/src/network/neuralnetwork.ts
import { lerp } from "../utils/utils";

export class NeuralNetwork {
  levels: Level[] = [];
  constructor(neurons: number[]) {
    for (let i = 0; i < neurons.length - 1; i++) {
      this.levels.push(new Level(neurons[i], neurons[i + 1]));
    }
  }

  static feedForward(inputs: number[], network: NeuralNetwork): boolean[] {
    let output = Level.feedForward(inputs, network.levels[0]);
    for (let i = 1; i < network.levels.length; i++) {
      output = Level.feedForward(output, network.levels[i]);
    }
    return output.map((o) => (o == 1 ? true : false));
  }
  static mutate(network: NeuralNetwork, amount = 0.1) {
    network.levels.forEach((level) => {
      for (let i = 0; i < level.biases.length; i++) {
        level.biases[i] = lerp(level.biases[i], Math.random() * 2 - 1, amount);
      }
      for (let i = 0; i < level.weights.length; i++) {
        for (let j = 0; j < level.weights[i].length; j++) {
          level.weights[i][j] = lerp(
            level.weights[i][j],
            Math.random() * 2 - 1,
            amount
          );
        }
      }
    });
  }
}

export class Level {
  inputs: number[] = [];
  outputs: number[] = [];
  weights: number[][] = [];
  biases: number[] = [];
  constructor(inputCount: number, outputCount: number) {
    this.inputs = new Array<number>(inputCount);
    this.outputs = new Array<number>(outputCount);
    this.biases = new Array<number>(outputCount);

    for (let i = 0; i < inputCount; i++) {
      this.weights[i] = new Array<number>(outputCount);
    }
    Level.randomize(this);
  }
  static randomize(level: Level) {
    for (let i = 0; i < level.inputs.length; i++) {
      for (let j = 0; j < level.outputs.length; j++) {
        level.weights[i][j] = Math.random() * 2 - 1;
      }
    }

    for (let i = 0; i < level.biases.length; i++) {
      level.biases[i] = Math.random() * 2 - 1;
    }
  }

  static feedForward(inputs: number[], level: Level): number[] {
    for (let i = 0; i < level.inputs.length; i++) {
      level.inputs[i] = inputs[i];
    }
    for (let i = 0; i < level.outputs.length; i++) {
      let sum = 0;
      for (let j = 0; j < level.inputs.length; j++) {
        sum += level.inputs[j] * level.weights[j][i];
      }
      level.outputs[i] = sum > level.biases[i] ? 1 : 0;
    }
    return level.outputs;
  }
}

File: raspi/python/neural_network/neural_network.py
import random
from typing import List


class Level:
    def __init__(self, input_count: int, output_count: int):
        self.inputs: List[float] = [0.0] * input_count
        self.outputs: List[float] = [0.0] * output_count
        self.biases: List[float] = [0.0] * output_count
        self.weights: List[List[float]] = [
            [0.0 for _ in range(output_count)] for _ in range(input_count)
        ]

        Level.randomize(self)

    @staticmethod
    def randomize(level: "Level"):
        for i in range(len(level.weights)):
            for j in range(len(level.weights[i])):
                level.weights[i][j] = random.uniform(-1, 1)

        for i in range(len(level.biases)):
            level.biases[i] = random.uniform(-1, 1)

    @staticmethod
    def feed_forward(inputs: List[float], level: "Level") -> List[float]:
        for i in range(len(level.inputs)):
            level.inputs[i] = inputs[i]

        for i in range(len(level.outputs)):
            total = 0.0
            for j in range(len(level.inputs)):
                total += level.inputs[j] * level.weights[j][i]
            level.outputs[i] = 1 if total > level.biases[i] else 0

        return level.outputs


class NeuralNetwork:
    def __init__(self):
        self.neurons: List[int] = [11, 16, 8, 4]
        self.levels: List[Level] = []
        for i in range(len(self.neurons) - 1):
            self.levels.append(Level(self.neurons[i], self.neurons[i + 1]))

    @staticmethod
    def feed_forward(inputs: List[float], network: "NeuralNetwork") -> List[bool]:
        output = Level.feed_forward(inputs, network.levels[0])
        for i in range(1, len(network.levels)):
            output = Level.feed_forward(output, network.levels[i])
        return [o == 1 for o in output]
